### PR TITLE
Add conflict resolution API to PP and CPP

### PR DESCRIPTION
**Pull request by RainbowMango**
**What type of PR is this?**

/kind api-change

**What this PR does / why we need it**:

This PR proposes the API changes for [Seamless Cluster Migration Proposal](https://github.com/karmada-io/karmada/pull/3780).

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE（Part of feature）
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
cc guys who are now reviewing the proposal.
@chaosi-zju @jwcesign @whitewindmills @zirain @yizhang-zen @chaunceyjiang @GitHubxsy

File: charts/karmada/_crds/bases/policy/policy.karmada.io_clusterpropagationpolicies.yaml [hunk 1]
```diff
@@ -49,6 +49,19 @@ spec:
...
+                description: "ConflictResolution declares how potential conflict should
+                  be handled when a resource that is being propagated already exists
+                  in the target cluster. \n It defaults to \"Abort\" which means stop
+                  propagating to avoid unexpected overwrites. The \"Overwrite\" might
+                  be useful when migrating legacy cluster resources to Karmada, in
+                  which case conflict is predictable and can be instructed to Karmada
+                  take over the resource by overwriting."
+                enum:
+                - Abort
+                - Overwrite
```
Comment (chaosi-zju) [inline charts/karmada/_crds/bases/policy/policy.karmada.io_clusterpropagationpolicies.yaml:63]: Should the initial letter be capitalized? The previous enum is `overwrite`. If we change to `Overwrite`, will it affect the use of some existing users.

Comment (RainbowMango) [reply charts/karmada/_crds/bases/policy/policy.karmada.io_clusterpropagationpolicies.yaml:63]: Yes, all enumerations should be in `CamelCase`.

> The previous enum is overwrite. If we change to Overwrite, will it affect the use of some existing users?

That's a good point! We might need to take care of it during implementation.

File: pkg/apis/policy/v1alpha1/propagation_types.go [hunk 2]
```diff
@@ -123,6 +123,19 @@ type PropagationSpec struct {
...
+
+	// ConflictResolution declares how potential conflict should be handled when
+	// a resource that is being propagated already exists in the target cluster.
+	//
+	// It defaults to "Abort" which means stop propagating to avoid unexpected
+	// overwrites. The "Overwrite" might be useful when migrating legacy cluster
+	// resources to Karmada, in which case conflict is predictable and can be
+	// instructed to Karmada take over the resource by overwriting.
+	//
+	// +kubebuilder:default="Abort"
```
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/apis/policy/v1alpha1/propagation_types.go:135]: As I mentioned at https://github.com/karmada-io/karmada/pull/3780#discussion_r1258229399, the `Ingore` might be ambiguous and can't tell the situation `ignore the conflict and continue propagation` and `skip the propagation action`.

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/apis/policy/v1alpha1/propagation_types.go:135]: I got the idea from Git that when resolving a conflict like a rebase, people can abort the previous operation.

Comment (whitewindmills) [inline pkg/apis/policy/v1alpha1/propagation_types.go:135]: the semantics of `Abort` in Git also include undoing the changes made, but we just stop propagating.

Comment (RainbowMango) [reply pkg/apis/policy/v1alpha1/propagation_types.go:135]: Yeah I get it. seems we have nothing to `undo` or revert, just stops. That's the meaning of `abort`.

Comment (chaunceyjiang) [reply pkg/apis/policy/v1alpha1/propagation_types.go:135]: > `Abort` means to terminate or cancel something, especially a planned action or process.

From chartGPT 😂

File: pkg/apis/work/v1alpha2/binding_types.go [hunk 3]
```diff
@@ -107,6 +107,19 @@ type ResourceBindingSpec struct {
...
+	//
+	// It defaults to "Abort" which means stop propagating to avoid unexpected
+	// overwrites. The "Overwrite" might be useful when migrating legacy cluster
+	// resources to Karmada, in which case conflict is predictable and can be
+	// instructed to Karmada take over the resource by overwriting.
+	//
+	// +kubebuilder:default="Abort"
+	// +kubebuilder:validation:Enum=Abort;Overwrite
+	// +optional
+	ConflictResolution policyv1alpha1.ConflictResolution `json:"conflictResolution,omitempty"`
```
Comment (chaosi-zju) [inline pkg/apis/work/v1alpha2/binding_types.go:122]: I think `ClusterResourceBinding` also need adding this field.

Comment (RainbowMango) [reply pkg/apis/work/v1alpha2/binding_types.go:122]: I think so, why not?  The cluster-scoped resources also need to be promoted, so both `ClusterPropagationPolicy` and `ClusterResourceBinding` need this.

**Comment by RainbowMango** [comment]
@whitewindmills @chaunceyjiang Do you have any further comments? If you have any other suggestions, please let me know.

File: pkg/apis/work/v1alpha2/binding_types.go (hunk 3 above)
Comment (chaosi-zju) [reply pkg/apis/work/v1alpha2/binding_types.go:122]: ok，the `ClusterResourceBingding` reused `ResourceBindingSpec`, I mistakenly thought you missed in `ClusterResourceBinding`.

**Review by chaunceyjiang — COMMENTED** [review]
/lgtm

**Review by whitewindmills — COMMENTED** [review]
/lgtm

**Comment by RainbowMango** [comment]
Thank you all!
@Garrybest would you like to take a look?

**Review by Garrybest — COMMENTED** [review]
Sorry for having this review late. This PR looks good to me.

/lgtm

**Pull request merged**